docs/decisions/0013-new-event-bus-producer-config.rst

```diff
@@ -0,0 +1,67 @@
+
+.. code-block:: python
+
+    { 'org.openedx.content_authoring.xblock.published.v1': [
+           {'topic': 'content-authoring-xblock-lifecycle', 'event_key_field': 'xblock_info.usage_key', 'enabled': True},
+           {'topic': 'content-authoring-xblock-published', 'event_key_field': 'xblock_info.usage_key', 'enabled': False},
+       ],
+       'org.openedx.content_authoring.xblock.deleted.v1': [
+           {'topic': 'content-authoring-xblock-lifecycle', 'event_key_field': 'xblock_info.usage_key', 'enabled': True},
+       ],
+   }
+
+While attempting to implement this for edx-platform, we came across some problems with using this structure. In particular, it results in ambiguity
+because maintainers can accidentally add something like
+``{'topic': 'content-authoring-xblock-lifecycle', 'event_key_field': 'xblock_info.usage_key', 'enabled': True}`` and
+``{'topic': 'content-authoring-xblock-lifecycle', 'event_key_field': 'xblock_info.usage_key', 'enabled': False}`` to the same event_type.
+Also, as written, this configuration requires maintainers to copy over the entire configuration to override it, which is non-obvious
+to people who may only be trying to enable/disable a single event while keeping everything else the same. Moreover, it's also non-obvious
+that enabling/disabling an existing event/topic pair requires reaching into the structure, searching for the dictionary with the correct topic, and modifying
+the existing object, which is awkward.
+
+This ADR aims to propose a new structure that will provide greater flexibility in using this configuration.
+
+Decision
+********
+
+The new EVENT_BUS_PRODUCER_CONFIG will have the following configuration format:
+
+.. code-block:: python
+
+   # .. setting_name: EVENT_BUS_PRODUCER_CONFIG
+   # .. setting_default: {}
+   # .. setting_description: Dictionary of event_types to dictionaries for topic related configuration.
+   #    Each topic configuration dictionary uses the topic as a key and contains a flag called `enabled`
+   #    denoting whether the event will be and `event_key_field` which is a period-delimited string path
+   #    to event data field to use as event key.
+   #    Note: The topic names should not include environment prefix as it will be dynamically added based on
+   #    EVENT_BUS_TOPIC_PREFIX setting.
+   EVENT_BUS_PRODUCER_CONFIG = {
+       'org.openedx.content_authoring.xblock.published.v1': {
+           'content-authoring-xblock-lifecycle': {'event_key_field': 'xblock_info.usage_key', 'enabled': False}
+           'content-authoring-xblock-published': {'event_key_field': 'xblock_info.usage_key', 'enabled': True}
+       },
+       'org.openedx.content_authoring.xblock.deleted.v1': {
+           'content-authoring-xblock-lifecycle': {'event_key_field': 'xblock_info.usage_key', 'enabled': True},
+       },
+   }
+
+In edx-platform, it will be added to the KEYS_WITH_MERGED_VALUES list to allow partial additions/overrides.
```

Comment:
Does `KEYS_WITH_MERGED_VALUES` work at all depths?

Reply:
@rgraber: I'm going to hold off on a deeper review until you confirm this works the way you want it to, and we confirm we have a working rollout plan.

Reply:
@robrap @rgraber In edx-platform, adding a key to KEYS_WITH_MERGED_VALUES doesn't actually ensure that any merging will occur. It just [ensures that the key is not overridden when loading the YAML](https://github.com/openedx/edx-platform/blob/e10ba22e51b10f7c48a314d209a366016017a35a/lms/envs/production.py#L74-L92). The merging itself must be done by hand later in production.py.

Consider, for example, REST_FRAMEWORK: by [adding it to KEYS_WITH_MERGED_VALUES](https://github.com/openedx/edx-platform/blob/e10ba22e51b10f7c48a314d209a366016017a35a/lms/envs/production.py#L86), we ensure that the existing value of REST_FRAMEWORK is not totally overridden by the YAML file. Then, later in production.py, we [update the existing value with the contents of ENV_TOKENS['REST_FRAMEWORK']](https://github.com/openedx/edx-platform/blob/e10ba22e51b10f7c48a314d209a366016017a35a/lms/envs/production.py#L1113-L1114), which comes from the YAML.

So, you could add EVENT_BUS_PRODUCER_CONFIG to KEYS_WTIH_MERGED_VALUES, and then later in production.py you could merge at whatever depth level you'd like.

(I'm not sure how other IDAs work.... it might be different).

Reply:
Do all yaml values come in as ENV_TOKENS? the comments make it seem like that's deprecated but I don't quite understand the alternative

Reply:
@robrap My vote would be to put together something that works for this particular case and then make a ticket to abstract it out to be of use to other tools.

Reply:
@rgraber Yes, ENV_TOKENS comes directly from the yaml, and it is currently the only way to read directly from the yaml when updating keys. I don't think the author of the comments on ENV_TOKENS understood that.

IDAs seem to do things differently. [When looping through DICT_UPDATE_KEYS (the IDA version of KEYS_WITH_MERGED_VALUES), they do a one-level-deep merge of yaml settings into existing settings](https://github.com/openedx/edx-cookiecutters/blob/df10fdadc365f41e81bf2429c8b68ab13be4cb3c/cookiecutter-django-ida/%7B%7Bcookiecutter.repo_name%7D%7D/%7B%7Bcookiecutter.project_name%7D%7D/settings/production.py#L17-L39). That would work fine in the case of keys like REST_FRAMEWORK, but wouldn't generalize to the arbitrary-depth merging you're interested in for EVENT_BUS_PRODUCER_CONFIG.

Reply:
@rgraber: I think ticketing sounds fine. Just clarifying that it isn’t just for different tools, but specifically for handling these producer configs in different services.

Reply:
This is in progress with https://github.com/openedx/edx-platform/pull/33269

Reply:
Ok. Marking this resolved.